homeassistant/components/smlight/sensor.py

```diff
@@ -0,0 +1,175 @@
+        device_class=SensorDeviceClass.TEMPERATURE,
+        native_unit_of_measurement=UnitOfTemperature.CELSIUS,
+        state_class=SensorStateClass.MEASUREMENT,
+        suggested_display_precision=1,
+        value_fn=lambda x: x.esp32_temp,
+    ),
+    SmSensorEntityDescription(
+        key="zigbee_temperature",
+        translation_key="zigbee_temperature",
+        device_class=SensorDeviceClass.TEMPERATURE,
+        native_unit_of_measurement=UnitOfTemperature.CELSIUS,
+        state_class=SensorStateClass.MEASUREMENT,
+        suggested_display_precision=1,
+        value_fn=lambda x: x.zb_temp,
+    ),
+    SmSensorEntityDescription(
+        key="ram_usage",
+        translation_key="ram_usage",
+        device_class=SensorDeviceClass.DATA_SIZE,
+        native_unit_of_measurement=UnitOfInformation.KILOBYTES,
+        entity_registry_enabled_default=False,
+        value_fn=lambda x: x.ram_usage,
+    ),
+    SmSensorEntityDescription(
+        key="fs_usage",
+        translation_key="fs_usage",
+        device_class=SensorDeviceClass.DATA_SIZE,
+        native_unit_of_measurement=UnitOfInformation.KILOBYTES,
+        entity_registry_enabled_default=False,
+        value_fn=lambda x: x.fs_used,
+    ),
+]
+UPTIME = [
+    SmSensorEntityDescription(
+        key="core_uptime",
+        translation_key="core_uptime",
+        device_class=SensorDeviceClass.TIMESTAMP,
+        entity_registry_enabled_default=False,
+        value_fn=lambda x: x.uptime,
+    ),
+    SmSensorEntityDescription(
+        key="socket_uptime",
+        translation_key="socket_uptime",
+        device_class=SensorDeviceClass.TIMESTAMP,
+        entity_registry_enabled_default=False,
+        value_fn=lambda x: x.socket_uptime,
+    ),
+]
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    entry: SmConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up SMLIGHT sensor based on a config entry."""
+    coordinator: SmDataUpdateCoordinator = entry.runtime_data
+
+    async_add_entities(
+        chain(
+            (SmSensorEntity(coordinator, description) for description in SENSORS),
+            (SmUptimeSensorEntity(coordinator, description) for description in UPTIME),
+        )
+    )
+
+
+class SmSensorEntity(SmEntity, SensorEntity):
+    """Representation of a slzb sensor."""
+
+    def __init__(
+        self,
+        coordinator: SmDataUpdateCoordinator,
+        description: SmSensorEntityDescription,
+    ) -> None:
+        """Initiate slzb sensor."""
+        super().__init__(coordinator)
+
+        self.entity_description: SmSensorEntityDescription = description
+        self._attr_unique_id = f"{coordinator.unique_id}_{description.key}"
+
+    @property
+    def native_value(self) -> float | datetime | None:
+        """Return the sensor value."""
+        return self.entity_description.value_fn(self.coordinator.data.sensors)
+
+
+class SmUptimeSensorEntity(SmSensorEntity):
```

Comment:
Why do we have a separate class for this?

Reply:
I split out the uptime sensors into a separate class based on suggestions from @zweckj in an earlier review. Mainly to separate the uptime handling/logic from plain sensors.

However I can revert this back to a single class similar to how I had it originally, if that would be preferred?

This was earlier revision of my sensor class:
https://github.com/home-assistant/core/blob/3403d235a98fb823da13aa9d19ce362baa811b0e/homeassistant/components/smlight/sensor.py#L109

Let me know if its preferred to merge it back into a single class and I will update accordingly.

Reply:
I have merged uptime sensors back in the main sensor class now.

Reply:
Note: this wasn't an action to do something, rather an exploratory question to know why this is chosen

Reply:
> Note: this wasn't an action to do something, rather an exploratory question to know why this is chosen

The review was postponed until changes were submitted. However, this PR is good with both single and separate classes, aligning with project practices. Can this be processed further with a single class? Or should it be separated?

Reply:
> Note: this wasn't an action to do something, rather an exploratory question to know why this is chosen

The review was postponed until changes were submitted. However, this PR is good with both single and separate classes, aligning with project practices. Can this be processed further with a single class? Or should it be separated?

Reply:
> Note: this wasn't an action to do something, rather an exploratory question to know why this is chosen

Because 
- The review was postponed until changes were submitted. 
- That was a part of "joostlek **requested changes**  [4 days ago]". 
Answer to the question is [here](https://github.com/home-assistant/core/pull/118675#discussion_r1698400920).

Just to clarify, this PR is good to go with both single and separate classes, aligning with project practices. 
Should it be separated again for your approval? Just let PR author know what is needed, i believe reverting back to separate classes is just a few minutes task.

Reply:
> Note: this wasn't an action to do something, rather an exploratory question to know why this is chosen
my 3rd version of this reply :)

changes were made because that was a part of "joostlek requested changes [4 days ago]". - misunderstanding probably. And, it blocks further commit checks as the status is still "changes requested" 
The answer to your question is here https://github.com/home-assistant/core/pull/118675#discussion_r1698400920.
Just to clarify, this PR is good to go with both single and separate classes, aligning with project practices.
Should it be separated again for your approval? Just please let the PR author know what is needed, i believe reverting back to separate classes is just a few minutes task.

Reply:
No it was more that the code looked like a lot, so instead of trying to understand what you are trying to do I wanted to consult what the code is doing so I then can use that info to check the code

Reply:
> No it was more that the code looked like a lot, so instead of trying to understand what you are trying to do I wanted to consult what the code is doing so I then can use that info to check the code

Coear

> No it was more that the code looked like a lot, so instead of trying to understand what you are trying to do I wanted to consult what the code is doing so I then can use that info to check the code

Got it, clear now, thanks, Joostlek. 
So to sum up, regarding the single class or separate classes question: 
what should be done:
1. to keep it as a single class (it is already as single), or 
2. to make it through multiple classes? (both options are valid and works). 

May it remain "as is" in the current PR and  users get it merged? Can you please review this  PR?

Reply:
> No it was more that the code looked like a lot, so instead of trying to understand what you are trying to do I wanted to consult what the code is doing so I then can use that info to check the code

Splitting the sensor class does add some complexity in itself. I not sure this outweighs the resulting simplification in the native_value property. Thus I could go either way on this. Regardless I have tried to clarify further the uptime handling code.

Reply:
But then my questions still stands, what extra calculation do we have to do for the uptime sensor?

Reply:
> But then my questions still stands, what extra calculation do we have to do for the uptime sensor?

There is no extra calculations for uptime just what you see there now.

The uptime is reported from the device in "Seconds since last boot", this is converted to a fixed datetime stamp, to minimise state changes of the sensor.  The deviation buffer allows any changes caused by clock skew etc to be ignored (we dont want to update the sensor state just because the calculated datetime varies by a few milliseconds). The HA frontend takes care of converting the uptime into a user friendly "3 hours" uptime.

This is very much specific of implementation of the sensor in HA, and the calculation also requires state only known within HA. So I think it makes more sense to have here, rather than in the library.